### PR TITLE
Minimal example to get debugging to work with iOS 17

### PR DESCRIPTION
**Pull request by yorickreum**
This pull request is addressing the issues https://github.com/nisargjhaveri/vscode-ios-debug/issues/18 and https://github.com/nisargjhaveri/vscode-ios-debug/issues/24 and contains a minimal example of what is needed to get the extension to work with Apple LLDB, and hence compatible with iOS 17. It requires the XCode 16 beta.
It's only using Apple LLDB and `xcrun devicectl` when really needed, the functions using ios-deploy are left as-is.

### CODE REVIEW

File: src/debugConfigProvider.ts [hunk 1]
```diff
@@ -162,7 +163,7 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
                     });
                 }
             }
-            else
+            else 
             {
                 platformPath = await targetCommand.deviceAppPath({
```
Comment (nisargjhaveri) [inline src/debugConfigProvider.ts:168]: Does this still work? Doesn't this also use `ios-deploy`?

File: src/debugConfigProvider.ts [hunk 2]
```diff
@@ -145,11 +146,11 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
-        {
-            let platformPath: string|void;
-            if (dbgConfig.iosRequest === "launch")
+        else if (target.type === "Device") 
             {
+            let platformPath: string|void;
+            if (dbgConfig.iosRequest === "launch") 
+                {
                 if (dbgConfig.iosInstallApp) {
                     platformPath = await targetCommand.deviceInstall({
```
Comment (nisargjhaveri) [inline src/debugConfigProvider.ts:155]: We'd also need to replace `deviceInstall`, right? Just noting, we can do this incrementally in a different PR as well.

File: src/debugConfigProvider.ts [hunk 3]
```diff
@@ -181,16 +182,37 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+
+                // attach to the app via LLDB
+                dbgConfig.processCreateCommands = (dbgConfig.postRunCommands instanceof Array) ? dbgConfig.processCreateCommands : [];
+                // LLDB `device` command needs Apple's beta LLDB set as backend for CodeLLDB
+                // e.g. set vscode://settings/lldb.library to /Applications/Xcode-beta.app/Contents/SharedFrameworks/LLDB.framework/Versions/A/LLDB
+                dbgConfig.processCreateCommands.push(`script lldb.debugger.HandleCommand("device select ${target.udid}")`);
+                // using detour with `script` command, as direct evocation causes CodeLLDB to crash at the moment
+                // probably due to a bug in CodeLLDB / unusal exit results from the `device` command
+                let processName = path.basename(platformPath).split('.')[0];
+                dbgConfig.processCreateCommands.push(`script lldb.debugger.HandleCommand("device process attach --name '${processName}'")`);
```
Comment (nisargjhaveri) [inline src/debugConfigProvider.ts:214]: How reliable attaching by name? Can the process have different name than binary? Should we try attaching to pid instead of name, if we can get pid from the process launch command above?

**Review by nisargjhaveri — COMMENTED** [review]
Thanks a lot for experimenting and creating this PR!

I'll try this out soon enough. In the meantime, I've added a few comments for the things that came to mind on first look.

File: src/debugConfigProvider.ts (hunk 2 above)
Comment (yorickreum) [reply src/debugConfigProvider.ts:155]: Installing the app still works with `ios-deploy`. Only getting the pid of a running process and spawning the debug server seems to require `DeveloperDiskImage.dmg` and hence is retired since iOS 17.
Of course long-term it would be sort of cleaner to use `xcrun devicectl` instead as well, but this will only work with XCode >= 15 installed, so at the time being `ios-deploy` actually provides the highest compatibility.

File: src/debugConfigProvider.ts (hunk 1 above)
Comment (yorickreum) [reply src/debugConfigProvider.ts:168]: At the moment, getting the platform path still works with `ios-deploy`.

One can also do it with `devicectl` of course, for running processes as the `runningProcesses.executable` of `xcrun devicectl device info apps  --json-output -`
and for all apps installed via `apps.url` of `xcrun devicectl device info apps --json-output -`.

File: src/debugConfigProvider.ts [hunk 4]
```diff
@@ -181,16 +182,37 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+                // launch the app via devicectl
+                await _execFile(
+                    'xcrun',
+                    ['devicectl', 'device', 'process', 'launch', '--device', target.udid, '--start-stopped', this.ensureBundleId(dbgConfig)],
+                );
+
+                // attach to the app via LLDB
+                dbgConfig.processCreateCommands = (dbgConfig.postRunCommands instanceof Array) ? dbgConfig.processCreateCommands : [];
+                // LLDB `device` command needs Apple's beta LLDB set as backend for CodeLLDB
+                // e.g. set vscode://settings/lldb.library to /Applications/Xcode-beta.app/Contents/SharedFrameworks/LLDB.framework/Versions/A/LLDB
```
[the remark this answers is not in the document]
Comment (yorickreum) [reply src/debugConfigProvider.ts:209]: I agree that it is an additional hurdle, but I can not think of a easy or clean way to programmatically set this, since we'd be messing with the settings of another app, and the path to Apple's LLDB library might also vary.
Maybe the cleanest option is to just send a clear alert to the user explaining what to do (whenever they are using iOS 17 or when `lldb` complains about `device` not being a command`).

File: src/debugConfigProvider.ts (hunk 3 above)
Comment (yorickreum) [reply src/debugConfigProvider.ts:214]: Not sure, pid might be cleaner. It gets returned as `process.processIdentifier` from `xcrun devicectl device process launch --json-output -`.

Cleanest option imo would be to use `lldb` also to start the process, but I couldn't figure out a way to do this with Apples `lldb device`.

**Comment by Staedyyy** [comment]
> This pull request is addressing the issues https://github.com/nisargjhaveri/vscode-ios-debug/issues/18 and https://github.com/nisargjhaveri/vscode-ios-debug/issues/24 and contains a minimal example of what is needed to get the extension to work with Apple LLDB, and hence compatible with iOS 17. It requires the XCode 16 beta.
> 
> It's only using Apple LLDB and `xcrun devicectl` when really needed, the functions using ios-deploy are left as-is.

**Comment by nisargjhaveri** [comment]
Apologies for very long delay. :/

I've made another attempt in #27. It does build somewhat on top of this PR, while taking care of logging. Also, attaches using pid instead of name for correctness.

Have a look if it makes sense?

**Comment by nisargjhaveri** [comment]
This was integrated and committed as 7870e842623e8920344f793e1e943035a755f89b and released in v0.4.1. Really appreciate the PR. Thanks! :)
